File: solutions/cpp/2140/solution.cpp
class Solution {
public:
    long long mostPoints2(vector<vector<int>>& questions) {
        vector<long long> memo(questions.size(), 0);
        return dp(0, memo, questions);
    }
    
    long long dp(int i, vector<long long> &memo, vector<vector<int>> &questions) {
        if (i >= questions.size()) {
            return 0;
        }
        
        if (memo[i] != 0) {
            return memo[i];
        }
        
        int j = i + questions[i][1] + 1;
        memo[i] = max(questions[i][0] + dp(j, memo, questions), dp(i + 1, memo, questions));
        return memo[i];
    }
    
    long long mostPoints(vector<vector<int>> &questions) {
        int n = questions.size();
        vector<long> dp(n + 1);
        
        for (int i = n - 1; i >= 0; i--) {
            int j = i + questions[i][1] + 1;
            dp[i] = max(questions[i][0] + dp[min(j, n)], dp[i + 1]);
        }
        
        return dp[0];
    }
};

File: solutions/golang/2140/solution.go
func mostPoints(questions [][]int) int64 {
	n := len(questions)
	dp := make([]int64, n+1)

	for i := n - 1; i >= 0; i-- {
		j := i + questions[i][1] + 1
		dp[i] = max(int64(questions[i][0])+dp[min(j, n)], dp[i+1])
	}

	return dp[0]
}

func mostPoints2(questions [][]int) int64 {
	memo := make([]int64, len(questions))
	for i := range memo {
		memo[i] = 0
	}

	return dp(questions, 0, memo)
}

func dp(questions [][]int, i int, memo []int64) int64 {
	if i >= len(questions) {
		return 0
	}

	if memo[i] != 0 {
		return memo[i]
	}

	j := i + questions[i][1] + 1
	memo[i] = max(int64(questions[i][0])+dp(questions, j, memo), dp(questions, i+1, memo))
	return memo[i]
}